src/trace/dma-trace.c

```diff
@@ -228,7 +228,8 @@ static int dma_trace_buffer_init(struct dma_trace_data *d)
 
 	if (!d || !d->dc.dmac) {
 		mtrace_printf(LOG_LEVEL_ERROR,
-			      "%s failed: no DMAC!", __func__);
+			      "dma_trace_buffer_init() failed: d=%p, DMAC=%p",
+			      d, d ? d->dc.dmac : NULL);
```

Comment:
@marc-hb, the pointer printing not going to help much as it will _always_ print:
`dma_trace_buffer_init() failed: d=NULL, DMAC=NULL`

What would be useful is something like this:
```c
	if (!d) {
		mtrace_printf(LOG_LEVEL_ERROR,
			"dma_trace_buffer_init() failed: no dma_trace_data!");
		return -ENODEV;
	}

	if (!d->dc.dmac) {
		mtrace_printf(LOG_LEVEL_ERROR,
			"dma_trace_buffer_init() failed: no DMAC!");
		return -ENODEV;
	}
```
Since we can not print string.

Reply:
> will always print dma_trace_buffer_init() failed: d=NULL, DMAC=NULL

I agree trying to print `d->dc.dmac` is useless. This was left over from an earlier version sorry.

However d can be not NULL because it has just been dereferenced so the system would have crashed earlier.

@jsarha could you fix this? I'd be delighted to abandon this PR...

Reply:
> > will always print dma_trace_buffer_init() failed: d=NULL, DMAC=NULL
> 
> I agree trying to print `d->dc.dmac` is useless. This was left over from an earlier version sorry.
> 
> However d can be not NULL because it has just been dereferenced so the system would have crashed earlier.

Haha, nice catch, that is bad as well. I'm sure in other places as well ;)

> 
> @jsarha could you fix this? I'd be delighted to abandon this PR...

Reply:
> However d can be not NULL because it has just been dereferenced so the system would have crashed earlier.

@marc-hb sorry, where has it been dereferenced before this check? In the caller? It's called from `dma_trace_enable()` from `ipc_dma_trace_config()` in the IPC3 case (not at all in IPC4 case?) and yes, `ipc_dma_trace_config()` did check, that `d != NULL` but `dma_trace_enable()` is an exported function, so you could argue... If you mean line 223, then it isn't a dereference, that's just an address calculation.

Reply:
@lyakh, line 223

Reply:
&(NULL)->dmatb is NULL dereference.

Reply:
> &(NULL)->dmatb is NULL dereference.

@lyakh, you are right, this is just address calculation, not a NULL dereference candidate.

Reply:
Thanks everyone, new version pushed. Printing only `d` now.

Reply:
I thought this still be NULL dereference. Thanks @marc-hb for a bit more C language help. I didn't know '&' make a big difference. Looks good.